### PR TITLE
[Compiler-v2] Treat unit as tuple in type constraints

### LINKED ISSUE #12764: [Bug][Compiler-v2] Unit type should not be allowed in struct
# 🐛 Bug

V2 type checker cannot generate error for the following code: 

```
module 0x8675309::M {
    struct Box<T> has drop { f: T }

    fun t0() {
        Box { f: () };
    }
}
```

V1 generates the following error:

```
error[E04004]: expected a single non-reference type
  ┌─ tests/move_check/typing/pack_unit.move:5:9
  │
5 │         Box { f: () };
  │         ^^^^^^^^^^^^^
  │         │        │
  │         │        Expected a single non-reference type, but found: '()'
  │         Invalid type argument
```

### PR DESCRIPTION
**Pull request by rahxephon89**
## Description

This PR closes #12764 by treating unit as tuple in type constraints. 

Note that the error message in the test `third_party/move/move-compiler-v2/tests/checking/typing/lambda.move` and `third_party/move/move-compiler-v2/tests/checking/typing/while_body_invalid.exp` is less clearer after this fix, which should be fixed.

## Type of Change
- [ ] New feature
- [x] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?

1) existing tests;
2) migrate the test `pack_unit` from V1

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: third_party/move/move-compiler-v2/tests/checking/typing/while_body_invalid.exp [hunk 1]
```diff
@@ -18,11 +18,11 @@ error: expected expression with no value but found `address`
 5 │         while (cond) { @0x0 };
   │                        ^^^^
 
-error: expected expression with no value but found `integer`
-  ┌─ tests/checking/typing/while_body_invalid.move:6:35
+error: tuple type `()` is not allowed
```
Comment (rahxephon89) [inline third_party/move/move-compiler-v2/tests/checking/typing/while_body_invalid.exp:21]: Error message needs to be improved.

File: third_party/move/move-compiler-v2/tests/checking/typing/loop_body_invalid.exp [hunk 2]
```diff
@@ -18,11 +18,11 @@ error: expected expression with no value but found `address`
 11 │         loop { @0x0 }
    │                ^^^^
 
-error: expected expression with no value but found `integer`
-   ┌─ tests/checking/typing/loop_body_invalid.move:15:27
+error: tuple type `()` is not allowed
```
Comment (rahxephon89) [inline third_party/move/move-compiler-v2/tests/checking/typing/loop_body_invalid.exp:21]: Error message needs to be improved.

File: third_party/move/move-compiler-v2/tests/checking/typing/lambda.exp [hunk 3]
```diff
@@ -38,8 +38,13 @@ error: reference type `&integer` is not allowed as a type argument (type was inf
    │
    = required by instantiating type parameter `T` of function `+`
 
-error: cannot pass `|&integer|integer` to a function which expects argument of type `|&integer|`
-   ┌─ tests/checking/typing/lambda.move:73:21
+error: tuple type `()` is not allowed as a type argument (type was inferred)
```
Comment (rahxephon89) [inline third_party/move/move-compiler-v2/tests/checking/typing/lambda.exp:41]: Error message needs to be improved.

File: third_party/move/move-model/src/ty.rs [hunk 4]
```diff
@@ -1718,7 +1718,7 @@ impl Substitution {
                     }
                 },
                 (Constraint::NoTuple, ty) => {
-                    if ty.is_tuple() {
+                    if ty.is_tuple() || ty.is_unit() {
```
Comment (vineethk) [inline third_party/move/move-model/src/ty.rs:1721]: `is_tuple` seems to be used in another file in a few places. Might be worth checking if a similar issue could exist in those places?

File: third_party/move/move-compiler-v2/tests/checking/typing/loop_body_invalid.exp (hunk 2 above)
Comment (vineethk) [reply third_party/move/move-compiler-v2/tests/checking/typing/loop_body_invalid.exp:21]: Please file a bug or add this case to an existing bug, if you have not already.

**Review by vineethk — APPROVED** [review]
LGTM, have some minor comments.

File: third_party/move/move-compiler-v2/tests/checking/typing/loop_body_invalid.exp (hunk 2 above)
Comment (rahxephon89) [reply third_party/move/move-compiler-v2/tests/checking/typing/loop_body_invalid.exp:21]: Will file the bug once the PR is landed

File: third_party/move/move-model/src/ty.rs (hunk 4 above)
Comment (rahxephon89) [reply third_party/move/move-model/src/ty.rs]: Thanks for pointing this out. I updated the implementation of `is_tuple` to include the unit type.

Comment (brmataptos) [reply third_party/move/move-model/src/ty.rs]: There are 3 uses in `exp_builder.rs` that you may want to examine to decide if this treatment is appropriate.

File: third_party/move/move-compiler-v2/tests/checking/typing/assign_wrong_arity.exp [hunk 5]
```diff
@@ -1,10 +1,20 @@
 
 Diagnostics:
-error: the left-hand side has 0 items but the right-hand side provided 3
-  ┌─ tests/checking/typing/assign_wrong_arity.move:7:9
+error: tuple type `()` is not allowed as a local variable type
+  ┌─ tests/checking/typing/assign_wrong_arity.move:5:13
   │
-7 │         x = (0, 1, 2);
-  │         ^
+5 │         let x;
```
[the remark this answers is not in the document]
Comment (rahxephon89) [reply third_party/move/move-compiler-v2/tests/checking/typing/assign_wrong_arity.exp:6]: Sure, we need to at lease point out that it is inferred as a unit type if no other type info is available.

[the remark this answers is not in the document]
Comment (rahxephon89) [reply third_party/move/move-compiler-v2/tests/checking/typing/assign_wrong_arity.exp:6]: I think the whole error message issue related to tuple needs to be tackled separately in other PRs.

**Review by brmataptos — APPROVED** [review]
Be sure to double-check the bad .exp changes and file an issue to fix them.

**Pull request merged**